File: GroceryList/Animation/CAPullToCloseTransitionManager.m
//
//  GLPullToCloseTransitionManager.m
//  GroceryList
//
//  Created by Hakon Hanesand on 3/29/15.

#import <pop/POP.h>
#import <ReactiveCocoa/ReactiveCocoa.h>

#import "GLPullToCloseTransitionManager.h"

#import "POPSpringAnimation+GLAdditions.h"
#import "POPAnimation+GLAnimation.h"

@implementation GLPullToCloseTransitionManager

- (NSTimeInterval)transitionDuration:(id<UIViewControllerContextTransitioning>)transitionContext {
    return 1.0;
}

- (void)animateTransition:(id<UIViewControllerContextTransitioning>)transitionContext {
    if (self.presenting) {
        [self animatePresentationWithTransitionContext:transitionContext];
    } else {
        [self animateDismissalWithTransitionContext:transitionContext];
    }
}


- (void)animatePresentationWithTransitionContext:(id<UIViewControllerContextTransitioning>)transitionContext {
    UIView *presentedControllerView = [transitionContext viewForKey:UITransitionContextToViewKey];
    UIView *containerView = [transitionContext containerView];
    
    //position the presented view below the screen
    presentedControllerView.center = translatePoint(containerView.center, 0, CGRectGetHeight(presentedControllerView.frame));
    
    [containerView addSubview:presentedControllerView];
    
    //animate the presented view up to cover the screen
    POPSpringAnimation *animation = [POPSpringAnimation animationWithPropertyNamed:kPOPLayerPositionY];
    animation.toValue = @(containerView.center.y);
    animation.springSpeed = 20;
    animation.springBounciness = 0;
    
    [[animation completionSignal] subscribeCompleted:^{
        [transitionContext completeTransition:![transitionContext transitionWasCancelled]];
    }];
    
    [presentedControllerView pop_addAnimation:animation forKey:@"presentPOPModal"];
}

- (void)animateDismissalWithTransitionContext:(id<UIViewControllerContextTransitioning>)transitionContext {
    UIView *presentedControllerView = [transitionContext viewForKey:UITransitionContextFromViewKey];
    UIView *containerView = [transitionContext containerView];
    
    //animate the view offscreen by moving it downwards
    POPSpringAnimation *animation = [POPSpringAnimation animationWithPropertyNamed:kPOPLayerPositionY];
    animation.toValue = @(translatePoint(containerView.center, 0, CGRectGetHeight(containerView.frame)).y);
    animation.springSpeed = 20;
    animation.springBounciness = 0;
    
    [[animation completionSignal] subscribeCompleted:^{
        [transitionContext completeTransition:![transitionContext transitionWasCancelled]];
    }];
    
    [presentedControllerView pop_addAnimation:animation forKey:@"dismissPOPModal"];
}

static inline CGPoint translatePoint(CGPoint point, CGFloat dx, CGFloat dy) {
    return CGPointMake(point.x + dx, point.y + dy);
}

@end

File: Carts/Animation/View Controller Presentation/CAPullToCloseTransitionManager.swift
//
//  CAPullToCloseTransitionManager.swift
//  Carts
//
//  Created by Hakon Hanesand on 4/26/15.
//  Copyright (c) 2015 Hakon Hanesand. All rights reserved.

import UIKit

class CAPullToCloseTransitionManager: NSObject, UIViewControllerAnimatedTransitioning, CAReversibleTransition {
    
    var presenting: Bool
    
    required override init() {
        presenting = true
        super.init()
    }
    
    func transitionDuration(transitionContext: UIViewControllerContextTransitioning) -> NSTimeInterval {
        return 1
    }
    
    func animateTransition(transitionContext: UIViewControllerContextTransitioning) {
        if self.presenting {
            self.animatePresentationWithTransitionContext(transitionContext)
        } else {
            self.animateDismissalWithTransitionContext(transitionContext)
        }
    }
    
    func animatePresentationWithTransitionContext(transitionContext: UIViewControllerContextTransitioning) {
        if let presentedControllerView = transitionContext.viewForKey(UITransitionContextToViewKey) {
            let containerView = transitionContext.containerView()
            
            presentedControllerView.center = self.translate(containerView.center, dx: 0, dy: CGRectGetHeight(presentedControllerView.frame))
            
            containerView.addSubview(presentedControllerView)
            
            let animation = POPSpringAnimation(propertyNamed: kPOPLayerPositionY)
            animation.toValue = containerView.center.y;
            animation.springSpeed = 20;
            animation.springBounciness = 0;
            
            animation.completionSignal().subscribeCompleted {
                transitionContext.completeTransition(!transitionContext.transitionWasCancelled())
            }
            
            presentedControllerView.pop_addAnimation(animation, forKey: "presentPOPModal")
        }
    }
    
    func animateDismissalWithTransitionContext(transitionContext: UIViewControllerContextTransitioning) {
        if let presentedControllerView = transitionContext.viewForKey(UITransitionContextFromViewKey) {
            let containerView = transitionContext.containerView()
            
            let animation = POPSpringAnimation(propertyNamed: kPOPLayerPositionY)
            animation.toValue = self.translate(containerView.center, dx: 0, dy: CGRectGetHeight(containerView.frame)).y;
            animation.springSpeed = 20;
            animation.springBounciness = 0;
            
            animation.completionSignal().subscribeCompleted {
                transitionContext.completeTransition(!transitionContext.transitionWasCancelled())
            }
            
            presentedControllerView.pop_addAnimation(animation, forKey: "dismissPOPModal")
            
        }
    }
    
    func translate(point: CGPoint, dx: CGFloat, dy: CGFloat) -> CGPoint {
        return CGPointMake(point.x + dx, point.y + dy);
    }
}
